src/Sarif.WorkItems/SarifWorkItemFiler.cs

```diff
@@ -0,0 +1,206 @@
+            // Now we initialize this work item model from the SARIF log file driving the
+            // current operations. As a first step, this code should extract all 
+            // source-controlled files from the log file. In an imminent implementation
+            // this information will be used to derive area path and owner information.
+            
+            workItemModel.Context.InitializeFromLog(sarifLog);
+
+            try
+            {
+                // Populate the work item with the target organization/repository information.
+                // In ADO, certain fields (such as the area path) will defaut to the 
+                // project name and so this information is used in at least that context. 
+                workItemModel.RepositoryOrProject = this.FilingClient.ProjectOrRepository;
+                workItemModel.OwnerOrAccount = this.FilingClient.AccountOrOrganization;
+
+                foreach (SarifWorkItemModelTransformer transformer in workItemModel.Context.Transformers)
+                {
+                    transformer.Transform(workItemModel);
+                }
+
+                this.FilingClient.FileWorkItems(new[] { workItemModel }).Wait();
+
+                // TODO: We need to process updated work item models to persist filing
+                //       details back to the input SARIF file, if that was specified.
+                //       This code should either return or persist the updated models
+                //       via a property, so that the file work items command can do
+                //       this work.
+                //
+                //       https://github.com/microsoft/sarif-sdk/issues/1774
+            }
+            catch (Exception ex)
+            {
+                Console.Error.WriteLine(ex.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Files work items from the results in a SARIF log file.
+        /// </summary>
+        /// <param name="projectUri">
+        /// The URI of the project in which the work items are to be filed.
+        /// </param>
+        /// <param name="workItemFilingModels">
+        /// Describes the work items to be filed.
+        /// </param>
+        /// <param name="personalAccessToken">
+        /// Specifes the personal access used to access the project. Default: null.
+        /// </param>
+        /// <returns>
+        /// The set of results that were filed as work items.
+        /// </returns>
+        internal async virtual Task<IEnumerable<WorkItemModel>> FileWorkItems(
+            Uri projectUri,
+            IList<WorkItemModel<SarifWorkItemContext>> workItemFilingModels,
+            string personalAccessToken = null)
+        {
+            if (projectUri == null) { throw new ArgumentNullException(nameof(projectUri)); }
+            if (workItemFilingModels == null) { throw new ArgumentNullException(nameof(workItemFilingModels)); }
+
+            FilingClient.Connect(personalAccessToken).Wait();
```

Comment:
>Wait [](start = 54, length = 4)

Instead of Wait(), can we await since the method is async?

Reply:
Yes, good catch, leftovers from debugging something.

---
In reply to: [381437615](https://github.com/microsoft/sarif-sdk/pull/1775#discussion_r381437615) [](ancestors = 381437615)